### PR TITLE
enhance issues#14

### PR DESCRIPTION
**Pull request by Yibo-Chen13**
Let PrettyPrintBlock support the reset of data types.

### CODE REVIEW

File: ut/utils.cpp [hunk 1]
```diff
@@ -111,15 +114,16 @@ bool doPrintValue<ColumnArray, void>(const ColumnRef & c, const size_t row, std:
 template <>
 bool doPrintValue<ColumnTuple, void>(const ColumnRef & c, const size_t row, std::ostream & ostr) {
     if (const auto & tupple_col = c->As<ColumnTuple>()) {
-        ostr << "(";
+        std::stringstream sstr;
```
Comment (yuzifeng1984) [inline ut/utils.cpp:117]: not sure if this is necessary. can we just use the single ostr?

Comment (Yibo-Chen13) [reply ut/utils.cpp:117]: Use the single ostr will make some format mistakes,like the following example:
```cpp
    cout << setw(10) << "(message1" << '\n';
    cout << setw(10) << "(";
    cout << "message2" << '\n';
```
This result is:
image

We have to output all the messages at once to make sure format correct.

File: ut/print_ut.cpp [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+    i128->Append(std::numeric_limits<Int128>::max() / 2);
+    i128->Append(std::numeric_limits<Int128>::min() / 2);
+    i128->Append(Int128{0});
+    i256->Append(std::numeric_limits<Int256>::max() / 2);
+    i256->Append(std::numeric_limits<Int256>::min() / 2);
+    i256->Append(Int128{0});
+    block.AppendColumn("int128", i128);
+    block.AppendColumn("int256", i256);
+
+    std::string correct_res = R"(
```
Comment (yuzifeng1984) [inline ut/print_ut.cpp:26]: nit: usually use 'expected' instead of 'correct' in unit test naming

File: ut/print_ut.cpp [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------+
+| Timeplus is a streaming-first data analytics platform. It provides powerful end-to-end capabilities to help teams process streaming and historical data quickly and intuitively, accessible for organizations of all sizes and industries. |       stream sql |
+|                                                                                                                                                                                                                 './,[]{}()*&^%$#%$#@!?<>": |         Timeplus |
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------+
+)";
+    std::stringstream sstr;
+    sstr << '\n' << PrettyPrintBlock{block};
+    EXPECT_TRUE(correct_res == sstr.str());
```
Comment (yuzifeng1984) [inline ut/print_ut.cpp:71]: can we use EXPECT_EQ macro? the difference is EXPECT_EQ will print both expected and actual value on assert failure so is easier for trouble shooting.

File: ut/print_ut.cpp [hunk 4]
```diff
@@ -0,0 +1,412 @@
+#include <gtest/gtest.h>
```
Comment (yuzifeng1984) [inline ut/print_ut.cpp:1]: Let's append these tests in utils_ut.cpp file instead of create a new file.

File: ut/utils.cpp [hunk 5]
```diff
@@ -203,6 +225,8 @@ std::ostream & operator<<(std::ostream & ostr, const ColumnValue& v) {
 
 }
 
+// map(low_cardinality(fixed_string(10)), array(nullable(uint64)))
+// {(key1, [1, 2, null, 4]), (key2, [null, 6, 7, 8])}
```
Comment (yuzifeng1984) [inline ut/utils.cpp:229]: can we remove this comment or add a more descriptive one to explain the function?

File: ut/utils.cpp [hunk 6]
```diff
@@ -213,14 +237,19 @@ std::ostream& operator<<(std::ostream & ostr, const PrettyPrintBlock & pretty_pr
...
     const auto horizontal_bar = '|';
     const auto cross = '+';
     const auto vertical_bar = '-';
 
     std::stringstream sstr;
     for (auto i = block.begin(); i != block.end(); ++i) {
         auto width = column_width[i.ColumnIndex()] = std::max(i.Type()->GetName().size(), i.Name().size());
+        for (size_t j = 0; j < block.GetRowCount(); ++j) {
+            std::stringstream tmp_stream;
+            printColumnValue(i.Column(),j,tmp_stream);
```
Comment (yuzifeng1984) [inline ut/utils.cpp:250]: add white space after comma.

**Review by yuzifeng1984 — APPROVED** [review]

**Pull request merged**